src/shared/install.c

```diff
@@ -1708,6 +1708,92 @@ static int install_info_discover_and_check(
         return install_info_may_process(ret ? *ret : NULL, paths, changes, n_changes);
 }
 
+int unit_file_verify_alias(const UnitFileInstallInfo *i, const char *dst, char **ret_dst) {
+        _cleanup_free_ char *inst1 = NULL, *inst2 = NULL, *dst2 = NULL;
+        _cleanup_free_ char *dir = NULL;
+        UnitNameFlags un1, un2;
+        UnitType t1, t2;
+        int r;
+
+        const char *path_alias = strrchr(dst, '/');
+        if (path_alias) {
+                dir = dirname_malloc(dst);
+                if (!dir)
+                        return log_oom();
+
+                if (!endswith(dir, ".wants") && !endswith(dir, ".requires"))
+                        return log_warning_errno(SYNTHETIC_ERRNO(EINVAL),
+                                                 "Invalid path \"%s\" in alias.", dir);
+
+                /* That's the name we want to use for verification. */
+                dst = path_alias + 1;
+        }
+
+        un1 = unit_name_to_instance(i->name, &inst1);
+        if (un1 < 0)
+                return log_error_errno(un1, "Unit name %s is not valid: %m", i->name);
+
+        un2 = unit_name_to_instance(dst, &inst2);
+        if (un2 < 0)
+                return log_warning_errno(un2, "Unit name %s is not valid: %m", dst);
+
+        t1 = unit_name_to_type(i->name);
+        assert(t1 >= 0);
+        t2 = unit_name_to_type(dst);
+        assert(t2 >= 0);
+
+        if (t2 != t1)
+                return log_warning_errno(SYNTHETIC_ERRNO(EINVAL),
+                                         "Unit %s alias target %s has wrong type.", i->name, dst);
+
+        if (un1 == UNIT_NAME_PLAIN) {
+                if (un2 != UNIT_NAME_PLAIN)
+                        return log_warning_errno(SYNTHETIC_ERRNO(EINVAL),
+                                                 "Unit %s cannot be aliased by template %s",
+                                                 i->name, dst);
+
+        } else if (un1 == UNIT_NAME_TEMPLATE) {
+                bool ok =
+                        un2 == UNIT_NAME_TEMPLATE ||
+                        (un2 == UNIT_NAME_INSTANCE && streq_ptr(i->default_instance, inst2)) ||
+                        (path_alias && un2 != UNIT_NAME_PLAIN);
```

Comment:
this is slightly stricter than what I did, but I guess that's fine.
This forbids Alias=template@my-%i but I think you are doing it on purpose..

Reply:
also, in the case of path_alias, un2 should be an instance unless the Wanting unit is also a template